sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -323,6 +324,18 @@ trait RuntimeReplaceable extends UnaryExpression with Unevaluable {
   // two `RuntimeReplaceable` are considered to be semantically equal if their "child" expressions
   // are semantically equal.
   override lazy val canonicalized: Expression = child.canonicalized
+
+  /**
+   * Only used to generate SQL representation of this expression.
+   *
+   * Implementations should override this with original parameters
+   */
+  protected def exprsReplaced: Seq[Expression]
+
+  override def sql: String = prettyName + exprsReplaced.map(_.sql).mkString("(", ", ", ")")
```

Comment:
another idea: maybe we can define a `mkString` method?
```
def exprsReplaced: Seq[Expression]

def mkString(childrenString: Seq[String]): String = childrenString.mkString("(", ", ", ")")

override def sql: String = prettyName + mkString(exprsReplaced.map(_.sql))
```

then in `util`
```
case r: RuntimeReplaceable => PrettyAttribute(r.mkString(r.exprsReplaced.map(toPrettySQL)), r.dataType)
```

Reply:
SGTM, this way is much better.  We should put the `prettyName` in `mkString` though.